### DISCUSSION TITLE
Reading 1 Gamecube Controller and passing the inputs to 4 Gamecube Consoles

### DISCUSSION THREAD

**Discussion by andrew171717** [NicoHood/Nintendo#69] [Q&A]
Any help would be appreciated. I'm trying to read one controller and pass the inputs to 4 consoles. Inputs pass just fine when calling  CGamecubeConsole.write for one console, but when I call it for multiple consoles, none of them receive input even though no error is returned. Testing with two console calls results in the same issue. If anyone has any suggestions for what I could try to fix this, please let me know. Any comments would be greatly appreciated. I've attached my code below.
```
#include "Nintendo.h"

// Define a Gamecube Controller and 4 Consoles
CGamecubeController GamecubeController(7);
CGamecubeConsole GamecubeConsole1(8);
CGamecubeConsole GamecubeConsole2(9);
CGamecubeConsole GamecubeConsole3(10);
CGamecubeConsole GamecubeConsole4(11);

// Pin definitions
#define pinLed LED_BUILTIN

void setup()
{
  // Set up debug led
  pinMode(pinLed, OUTPUT);

  // Start debug serial
  Serial.begin(19200);
}


void loop()
{
  if (GamecubeController.read()){
    if (!GamecubeConsole1.write(GamecubeController))){
      Serial.println(F("Error writing Gamecube Console 1"));
    }
    if (!GamecubeConsole2.write(GamecubeController))){
      Serial.println(F("Error writing Gamecube Console 2"));
    }
    if (!GamecubeConsole3.write(GamecubeController))){
      Serial.println(F("Error writing Gamecube Console 3"));
    }
    if (!GamecubeConsole4.write(GamecubeController))){
      Serial.println(F("Error writing Gamecube Console 4"));
    }
  }
  else
  {
    Serial.println(F("Error reading Gamecube controller."));
    digitalWrite(pinLed, HIGH);
    delay(1000);
  }
  digitalWrite(pinLed, LOW);
}
```

**Comment by andrew171717** [comment]
I could be wrong, but I think it's coming down to clock speed. My current Arduino runs at 16MHz. I've purchased an Arduino Due that runs at 80MHz. I'll report back once I've had a chance to test it tomorrow.

**Comment by Skuzee** [comment]
Console.write is a blocking function. That is, when called, it will halt the program and monitor the data line waiting for the wii to send a command before it continues. It might be likely that the consoles do not request the data in the perfect order. So it's stuck on console1 when the others are trying to request data.
  
If I remember correctly, it takes about 350 micro seconds to send the data packet. So sending to 4 consoles shouldn't be too problematic.
  
Lastly, this library timing has hard coded assembly timing, so it only works on 16mhz avr processors.
  
I known someone who made it work on a teensy using C, but I don't think the processor speed is the biggest problem.  
https://github.com/DogeSSBM/DogeBawx 
 
Perhaps there's a way you could edit the library to monitor all 4 console data lines at the same time, instead of individually.

**Comment by andrew171717** [comment]
@Skuzee Thanks for the response. 

That all makes sense. I'll see what I can do with the library's code. I'm not super familiar with assembly, so this will no doubt take me a bit.

**Reply by Skuzee** [reply]
Only the timing of sending the raw data pulses is assembly. The console write function should be c or c++

**Reply by andrew171717** [reply]
I've been messing with the functions. From my digging, the assembly that will need to be modified to make this work is in gc_n64_get and probably gc_n64_send. I've made progress c side creating clones of these functions called gc_n64_get_multi and gc_n64_send_multi. I've passed all the values I would need to those functions and passed to the assembly. Unfortunately, that's where my progress pretty much ends. Every time I try to modify the assembly to communicate with more than one console, the timing is so off that no inputs are registered on the either console and I'm not sure how to address that.

I'm starting think it'd be easier and more reliable if I setup 4 Arduinos and passed the controller data from one to the rest and send the controller data to each console from their own Arduino. I really don't like the idea of throwing so many boards at the problem, but I don't think it's feasible for me to understand the timings involved in these functions let alone know how to actually modify them to work with multiple consoles.

*Sigh* I really want to get this to work with one board, so I'm going to keep trying for now. Any suggestions of other resources or ideas would be really appreciated. Thanks again for your help

Edit: I did see this old closed issue https://github.com/NicoHood/Nintendo/issues/21 So that's unfortunate.

**Comment by andrew171717** [comment]
Ok, I was able to get it working. I went the route of 5 Arduinos. 1xMega (has 4 serial connections) and 4xUnos (1 serial connection each). I went ahead and created a github repo for the project and uploaded everything.

**Comment by andrew171717** [comment]
I have a working final project now. See here: https://www.drewworks.dev/shop
